test/fake_pipelineprov.c

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright 2020-2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <stdio.h>
+#include <openssl/core.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/proverr.h>
+#include "prov/provider_ctx.h"
+#include "prov/implementations.h"
+#include "prov/providercommon.h"
+#include "prov/ciphercommon.h"
+#include "prov/ciphercommon_gcm.h"
+#include "cipher_aes_gcm.h"
+#include "prov/names.h"
+#include "prov/provider_util.h"
+#include "internal/param_names.h"
+#include "testutil.h"
+#include "fake_pipelineprov.h"
+
+/*
+ * This file provides a fake provider that implements a pipeline cipher
+ * for AES GCM.
+ */
+
+typedef struct prov_gcm_pipeline_ctx_st {
+    void *provctx;
+    size_t keybits;
+    size_t numpipes;
+    PROV_AES_GCM_CTX *cipher_ctxs;
+    void *hw;
+} PIPE_CTX;
+
+const PROV_GCM_HW *ossl_prov_aes_hw_gcm(size_t keybits);
+static void *aes_gcm_pipeline_newctx(void *provctx, size_t keybits)
+{
+    PIPE_CTX *ctx;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx != NULL) {
+        ctx->provctx = provctx;
+        ctx->keybits = keybits;
+        ctx->numpipes = 0;
+        ctx->cipher_ctxs = NULL;
+        ctx->hw = (void *) ossl_prov_aes_hw_gcm(ctx->keybits);
+    }
+    return ctx;
+}
+
+static OSSL_FUNC_cipher_freectx_fn aes_gcm_pipeline_freectx;
+static void aes_gcm_pipeline_freectx(void *vctx)
+{
+    PIPE_CTX *ctx = (PIPE_CTX *)vctx;
+
+    OPENSSL_clear_free(ctx, sizeof(*ctx));
+}
+
+OSSL_FUNC_cipher_pipeline_encrypt_init_fn test_gcm_pipeline_einit;
+OSSL_FUNC_cipher_pipeline_decrypt_init_fn test_gcm_pipeline_dinit;
+OSSL_FUNC_cipher_pipeline_update_fn test_gcm_pipeline_update;
+OSSL_FUNC_cipher_pipeline_final_fn test_gcm_pipeline_final;
+OSSL_FUNC_cipher_gettable_ctx_params_fn test_cipher_aead_gettable_ctx_pipeline_params;
+OSSL_FUNC_cipher_get_ctx_params_fn test_gcm_get_ctx_pipeline_params;
+OSSL_FUNC_cipher_settable_ctx_params_fn test_cipher_aead_settable_ctx_pipeline_params;
+OSSL_FUNC_cipher_set_ctx_params_fn test_gcm_set_ctx_pipeline_params;
+
+int test_gcm_pipeline_einit(void *vctx,
+                            const unsigned char *key, size_t keylen,
+                            size_t numpipes, const unsigned char **iv,
+                            size_t ivlen, const OSSL_PARAM params[])
+{
+    PIPE_CTX *ctx = (PIPE_CTX *)vctx;
+
+    ctx->cipher_ctxs = OPENSSL_zalloc(numpipes * sizeof(PROV_AES_GCM_CTX));
+    if (ctx->cipher_ctxs == NULL)
+        return 0;
+    ctx->numpipes = numpipes;
+    for (size_t i = 0; i < numpipes; i++) {
+        ossl_gcm_initctx(ctx->provctx, (PROV_GCM_CTX *)(ctx->cipher_ctxs + i),
```

Comment:
Right. Do you suggest changing it that way?

Reply:
I think it will be easier to maintain doing it that way because we will just use public API functions - rather than relying on internal code which code change. So it would be better to change it IMO.

Reply:
Reworked fake provider to use public API